File: Yash/C++/DSA/3. LinkedList/insertInBetween.cpp
#include <iostream>
using namespace std;

class Node{
    public:
    int data;
    Node *next;

    Node(int data)
    {
        this->data = data;
        next = nullptr;
    }
};

Node *createLinkedList()
{
    Node *n1 = new Node(10);
    Node *n2 = new Node(20);
    Node *n3 = new Node(30);
    Node *n4 = new Node(50);
    Node *n5 = new Node(60);
    
    n1->next = n2;
    n2->next = n3;
    n3->next = n4;
    n4->next = n5;

    return n1;
}

void printLinkedList(Node *head)
{
    Node *temp = head;
    while(temp!= nullptr)
    {
        cout<<temp->data<<endl;
        temp= temp->next;
    }
}

Node *insertAtIndex(Node*head, int pos, int data)
{
    Node *currentNode = new Node(data);
    Node *previousNode = nullptr;
    Node *nextNode = nullptr;
    Node *temp = head;
    for(int i = 0; i<(pos-1); i++)
    {
        previousNode = temp;
        temp = temp->next;
    }
    nextNode = previousNode->next;
    previousNode->next = currentNode;
    currentNode->next = nextNode;

    return head;
}

int main()
{
    int data, pos;
    Node *head = createLinkedList();
    cout<<"Enter the data to be inserted : ";
    cin>>data;
    cout<<"\nEnter the position : ";
    cin>>pos;

    cout<<"The older LinkedList is "<<endl;
    printLinkedList(head);

    cout<<"The new LinkedList is "<<endl;
    Node *modified_LL = insertAtIndex(head,pos, data);
    printLinkedList(modified_LL);
}

File: Yash/Java/5. LinkedList/InsertInBetween.java
package Linked_revision;
import java.util.Scanner;
public class InsertInBetween {
	public static Node<Integer> createLinkedList()
	{
		Node<Integer> n1 = new Node<>(10);
		Node<Integer> n2 = new Node<>(20);
		Node<Integer> n3 = new Node<>(40);
		Node<Integer> n4 = new Node<>(50);
		n1.next = n2;
		n2.next = n3;
		n3.next = n4;
		return n1;
	}
	
	public static void printLinkedList(Node<Integer> head)
	{
		while(head != null)
		{
			System.out.println(head.data);
			head = head.next;
		}
	}
	public static void insertLinkedList(Node<Integer> head)
	{
		Scanner s = new Scanner(System.in);
		System.out.print("Enter the element to be inserted : ");
		int data = s.nextInt();
		System.out.print("Enter the position the inserted : ");
		int pos = s.nextInt();
		Node<Integer> newNode = new Node<>(data);
		Node<Integer> previousNode = null;
		Node<Integer> nextNode = null;
		for(int i = 0; i<pos-1; i++)
		{
			
			previousNode = head.next;
			nextNode = head.next.next;
		}
		//insertion-part
		newNode.next = nextNode;
		previousNode.next = newNode;
	}
	
	public static void main(String[] args)
	{
		Node<Integer> head = createLinkedList();
		System.out.print("Before inserion : " );
		printLinkedList(head);
		insertLinkedList(head);
		System.out.print("After insertion : " );
		printLinkedList(head);
	}
}
